src/wh_server_crypto.c

```diff
@@ -649,6 +647,79 @@ int wh_Server_HandleCryptoRequest(whServerContext* server,
         }
         break;
 #endif /* !WC_NO_RNG */
+#ifdef WOLFSSL_CMAC
+    case WC_ALGO_TYPE_CMAC:
+        /* in, out and key are after the fixed size fields */
+        in = (uint8_t*)(&packet->cmacReq + 1);
+        key = in + packet->cmacReq.inSz;
+        out = (uint8_t*)(&packet->cmacRes + 1);
+        /* do oneshot if all fields are present */
+        if (packet->cmacReq.inSz != 0 && packet->cmacReq.keySz != 0 &&
+            packet->cmacReq.outSz != 0) {
+            field = packet->cmacReq.outSz;
+            ret = wc_AesCmacGenerate_ex(server->crypto->cmac, out, &field, in,
+                packet->cmacReq.inSz, key, packet->cmacReq.keySz, NULL,
+                server->crypto->devId);
+            packet->cmacRes.outSz = field;
+        }
+        else {
+            /* do each operation based on which fields are set */
+            /* set key if present, otherwise load the struct from keyId */
+            if (packet->cmacReq.keySz != 0) {
+                /* initialize cmac with key and type */
+                ret = wc_InitCmac_ex(server->crypto->cmac, key,
+                    packet->cmacReq.keySz, packet->cmacReq.type, NULL, NULL,
+                    server->crypto->devId);
+            }
+            else {
+                field = sizeof(server->crypto->cmac);
```

Comment:
Did you mean server->crypto->cmac->aes.key?  This is the key schedule, so it's different than the normal key and you'll need to invoke AesSetKey on it anyway.  Recommend that you read the key into a temp buffer, invoke set key, and then carry on.

Reply:
no we can just move the cmac struct into and out of the cache when we need to pause and resume.